### PR TITLE
Use best block state to craft message relay transactions

### LINKED ISSUE #432: Use best chain state to make decisions about transactions in messages relay
Right now (in #399) there's custom strategy for the delivery race. It makes decision using data received from best **finalized block** state. But since transactions are updating **best block** state, we need to use data from best block state.

### PR DESCRIPTION
**Pull request by svyatonik**
closes #432
should be retested after #493

### CODE REVIEW

File: relays/messages-relay/src/message_lane_loop.rs [hunk 1]
```diff
@@ -750,31 +756,36 @@ pub(crate) mod tests {
...
-				if data.source_state.best_peer.0 < data.target_state.best_self.0 {
-					data.source_state.best_peer = data.target_state.best_self;
+				if data.source_state.best_finalized_peer_at_best_self.0 < data.target_state.best_finalized_self.0 {
+					data.source_state.best_finalized_peer_at_best_self = data.target_state.best_self;
 				}
 				// if target has received all messages => increase target block so that confirmations may be sent
 				if data.target_latest_received_nonce == 10 {
 					data.target_state.best_self =
 						HeaderId(data.source_state.best_self.0 + 1, data.source_state.best_self.0 + 1);
+					data.target_state.best_finalized_self = data.target_state.best_finalized_self;
```
Comment (tomusdrw) [inline relays/messages-relay/src/message_lane_loop.rs:788]: ?
```suggestion
					data.target_state.best_finalized_self = data.source_tate.best_finalized_self;
```

**Review by HCastano — APPROVED** [review]

File: relays/messages-relay/src/message_race_loop.rs [hunk 2]
```diff
@@ -358,10 +367,12 @@ pub async fn run<P: MessageRace, SC: SourceClient<P>>(
...
 				let at_block = race_state
-					.source_state
+					.best_finalized_source_header_id_at_source
 					.as_ref()
-					.expect("source_nonces_required is only true when source_state is Some; qed")
-					.best_self
+					.expect(
+						"source_nonces_required is only true when\
+						best_finalized_source_header_id_at_source is Some; qed",
+					)
```
[the remark this answers is not in the document]
Comment (svyatonik) [reply relays/messages-relay/src/message_race_loop.rs:375]: In the code that sets `source_nonces_required` to `true`

File: relays/messages-relay/src/message_race_loop.rs [hunk 3]
```diff
@@ -388,10 +399,9 @@ pub async fn run<P: MessageRace, SC: SourceClient<P>>(
 			if target_nonces_required {
 				log::debug!(target: "bridge", "Asking {} about message nonces", P::target_name());
 				let at_block = race_state
-					.target_state
+					.best_target_header_id
 					.as_ref()
-					.expect("target_nonces_required is only true when target_state is Some; qed")
-					.best_self
+					.expect("target_nonces_required is only true when best_target_header_id is Some; qed")
```
[the remark this answers is not in the document]
Comment (svyatonik) [reply relays/messages-relay/src/message_race_loop.rs:404]: In the code that sets `target_nonces_required` to true

**Review by tomusdrw — APPROVED** [review]

File: relays/messages-relay/src/message_race_loop.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (HCastano) [reply relays/messages-relay/src/message_race_loop.rs:375]: Yeah, I guess looking back it's a bit obvious where it is, but I missed it yesterday.

**Pull request merged**